StructuredConcurrency.md

```diff
@@ -0,0 +1,289 @@
+both timely, safe and efficient in a wide variety of situations? Ideally we'd
+like tight numerical loops to be cancellable as well as IO. And we want all
+this without the performance penalty of inserting extra checks or safe points
+into loop code.
+
+### The challenge of preemptive cancellation
+
+At first sight one might hope to treat preemptive cancellation somewhat like
+`InterruptException`: wake the task, deliver a signal to its thread to generate
+a `CanceledException` which then unwinds the stack, running regular user
+cleanup code.
+
+The key difficulty here is that arbitrary preemptive cancellation can occur in
+any location with no syntactic hint in the source. Others [have
+claimed](https://github.com/golang/go/issues/29011#issuecomment-443441031) that
+this makes arbitrary cancellation an impossible problem for user code. The
+standard compromise is to make only a core set of operations (including IO)
+cancellable. This is the solution offered in
+[Python Trio checkpoints](https://trio.readthedocs.io/en/stable/reference-core.html#checkpoints),
+libdill's family of IO functions and in pthreads (see [pthread\_cancel](http://man7.org/linux/man-pages/man3/pthread_cancel.3.html)
+and [pthreads cancellation points](http://man7.org/linux/man-pages/man7/pthreads.7.html)).
+In contrast, consider the failed preemptive cancellation APIs
+[Java `Thread.stop`](https://docs.oracle.com/javase/1.5.0/docs/guide/misc/threadPrimitiveDeprecation.html)
+Windows API
+[`TerminateThread`](https://devblogs.microsoft.com/oldnewthing/?p=91811),
+both of which were found to be fundamentally non-robust.
+
+Let's consider the ways in which current julia code can be non-robust in the
+face of `InterruptException`. A particular difficulty occurs in resource
+acquisition. Consider this snippet from task.jl:
+
+```julia
+lock(t.donenotify)
+# < What if InterruptException is thrown here?
+try
+    while !istaskdone(t)
+        wait(t.donenotify)
+    end
+finally
+    unlock(t.donenotify)
+end
+```
+
+In Julia we have the escape hatch `disable_sigint` (`jl_sigatomic_begin` in the
+runtime) for deferring `InterruptException`, but most code doesn't consider or
+use this which makes user resource handling broken by default.
+
+So it's fairly clear that arbitrary cancellation without cleanup is a
+non-starter and that arbitrary cancellation with cleanup is difficult. But that
+leaves us in a difficult situation: how do we allow for cancellation of
+expensive numerical operations? Are there options for cancellation of numerical
+loops with a semantic which can be understood by users? The Go people seem to
+consider that arbitrary *preemption* is workable, but can arbitrary
+cancellation be made to work with the right language and library features?
+
+#### Runtime technicalities for preemption
+
+On a technical level, our runtime situation in julia-1.3 is very similar to
+Go where preemption is cooperative and a rouge goroutine can sometimes wedge
```

Comment:
Typo: `rouge` -> `rogue`

Reply:
Hi Arch, thanks; will be fixed in the next revision. If you do have time to provide some perspective, please head on over to

https://github.com/JuliaLang/julia/issues/33248

where I am pulling on a lot of disparate threads and undoubtedly need all the help I can get from people who actually know things :-)